File: go/fetch/response.go
package fetch

import "net/http"

// FetchResponse wraps an HTTP response with parsed body data and metadata.
type FetchResponse[T any] struct {
	// StatusCode is the HTTP status code.
	StatusCode int
	// Status is the full status string (e.g. "200 OK").
	Status string
	// Headers are the response headers.
	Headers http.Header
	// Data is the parsed response body. For JSON responses this is the decoded value;
	// for non-JSON it is the zero value of T.
	Data T
	// IsJSON indicates whether the response body was parsed as JSON.
	IsJSON bool
	// BodyRaw is the raw response body bytes.
	BodyRaw []byte
	// OK is true if the status code is in the 2xx range.
	OK bool
}

File: rust/fetch/src/response.rs
//! Response types for the smooai-fetch client.

use std::collections::HashMap;

/// A generic response structure that holds parsed data and metadata.
///
/// Analogous to the TypeScript `ResponseWithBody<T>`.
#[derive(Debug, Clone)]
pub struct FetchResponse<T> {
    /// HTTP status code.
    pub status: u16,
    /// HTTP status text (reason phrase).
    pub status_text: String,
    /// Whether the response was successful (status 200-299).
    pub ok: bool,
    /// Response headers.
    pub headers: HashMap<String, String>,
    /// Whether the response body was parsed as JSON.
    pub is_json: bool,
    /// The raw response body as a string.
    pub body: String,
    /// The parsed/deserialized response data, if available.
    pub data: Option<T>,
}

impl<T> FetchResponse<T> {
    /// Create a new FetchResponse from raw components.
    pub fn new(
        status: u16,
        status_text: String,
        headers: HashMap<String, String>,
        body: String,
        is_json: bool,
        data: Option<T>,
    ) -> Self {
        let ok = (200..300).contains(&(status as u32));
        Self {
            status,
            status_text,
            ok,
            headers,
            is_json,
            body,
            data,
        }
    }
}
